{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workshop 2 - Unix Shell Basics\n",
    "\n",
    "This workshop is designed to build your skills with one of the most popular forms of the unix command-line, `Bash`.  \n",
    "\n",
    "Before you begin this workshop you should have familiarised yourself with chapters 1-20 of the [interactive guide](http://bc2023.bioinformatics.guide/lessons/).  \n",
    "\n",
    "At the end of this workshop you should be able to;\n",
    "\n",
    "1. Use many new unix commands including `wc`, `head`, `tail`, `grep`, `sort` and `cut`.\n",
    "2. Chain commands together with the pipe `|` operator.\n",
    "3. Redirect output of commands to a file with the `>` operator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "nbgrader": {
     "grade": false,
     "grade_id": "important",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "source": [
    "# IMPORTANT\n",
    "\n",
    "> ## Run the Setup Code \n",
    "\n",
    "In order for this notebook to work properly you need to **run the cell below before doing anything else**. This will load custom functions and settings required to make the self assessment exercises work. \n",
    "\n",
    "If you restart your kernel you will also need to rerun the setup code \n",
    "\n",
    "> ## Don't use the `cd` command \n",
    "\n",
    "The answers to all self assessment exercises assume that you don't change your directory from the default.  You shouldn't ever need to use the `cd` command to answer an exercise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "nbgrader": {
     "grade": false,
     "grade_id": "setup",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Setup Done\n"
     ]
    }
   ],
   "source": [
    "wget -q https://www.dropbox.com/s/4tv0mebn25vy1cq/setup.sh?dl=0 -O setup.sh\n",
    "source setup.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Background\n",
    "\n",
    "In this tutorial we will work with the genome of [_Helicobacter pylori_](https://en.wikipedia.org/wiki/Helicobacter_pylori).  Australian scientist [Barry Marshal](https://en.wikipedia.org/wiki/Barry_Marshall) famously infected himself with this bacterium in order to prove that it was the causative agent of gastric ulcers.  He and [Robin Warren](https://en.wikipedia.org/wiki/Robin_Warren) won the 2005 Nobel prize for their discovery.\n",
    "\n",
    "<a title=\"By Yutaka Tsutsumi, M.D. Professor Department of Pathology Fujita Health University School of Medicine [Copyrighted free use], via Wikimedia Commons\" href=\"https://commons.wikimedia.org/wiki/File%3AEMpylori.jpg\"><img width=\"256\" alt=\"EMpylori\" src=\"https://upload.wikimedia.org/wikipedia/commons/d/d6/EMpylori.jpg\" style=\"float: right;margin-right: 7px;margin-top: 7px;\"/></a>\n",
    "\n",
    "The genome of _H. pylori_ is available for download from all major sequence databases.  As an exercise try finding the genome page for this organism. \n",
    "\n",
    "1. Visit [https://www.ncbi.nlm.nih.gov/genome](https://www.ncbi.nlm.nih.gov/genome)\n",
    "2. Type _Helicobacter pylori_ in the search box\n",
    "\n",
    "The webpage that appears is a typical genome entry.  Not all organisms have sequenced genomes but those that do typically have a dedicated page like this with links to key genomic resources. Notice that there are links to the genome itself, the sequences of all proteins as well as a bunch of files under the heading of **genome annotation**.\n",
    "\n",
    "In this tutorial we will work with the genome annotation for _H. pylori_\n",
    "\n",
    "![H. pylori Genome Page](h_pylori_ncbi.png)\n",
    "\n",
    "As a side note you will probably notice that the genome page for this organism indicates that there are a large number of available genome sequences (755 at last count).  Since this is a well studied bacterium the genomes of many strains have been sequenced. NCBI page defaults to showing the genome of the reference strain but the genomes of all the others are also available.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the `ls` command to list the files in current directory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You should see a file called `h_pylori.gff`.  This file is in a text based tabular format called `gff` (genome features file). The complete listing of positions and descriptions of genes and other features in a genome is called a genome annotation. The `gff` format is a common way of storing annotations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 1\n",
    "\n",
    "The `h_pylori.gff` file contains entries for many thousands of genes. One command which is very useful for peeking inside a large file like this is the `head` command. \n",
    "\n",
    "The `head` command shows the top few lines of a file.  Try it out by entering the command shown below in a [test cell](http://bc2023.bioinformatics.guide/Guide#create-test-cells).\n",
    "\n",
    "```\n",
    "head h_pylori.gff\n",
    "```\n",
    "\n",
    "As you might expect there is also a command called `tail` which shows the last few lines of a file.  Try it out like this\n",
    "\n",
    "```\n",
    "tail h_pylori.gff\n",
    "```\n",
    "\n",
    "Now look at the help for the `head` command like this\n",
    "\n",
    "```\n",
    "head --help\n",
    "```\n",
    "\n",
    "**Your Task:** Write code to display the first 5 lines of `h_pylori.gff`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e1_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e1_answer(){\n",
    "### BEGIN SOLUTION\n",
    "head -n 5 h_pylori.gff\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e1_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test_e1: command not found\n"
     ]
    },
    {
     "ename": "",
     "evalue": "127",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "test_e1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GFF Format\n",
    "\n",
    "Examine the entries in `h_pylori.gff`. The [gff](http://www.ensembl.org/info/website/upload/gff.html) format is a tabular text based format that consists of up to $9$ columns;\n",
    "\n",
    "The columns (in order) are;\n",
    "\n",
    "1. **seqname** - name of the chromosome. \n",
    "2. source - name of the program or project that generated this feature\n",
    "3. **feature** - feature type name, e.g. Gene, Exon, rRNA\n",
    "4. **start** - Start position of the feature, with sequence numbering starting at 1.\n",
    "5. **end** - End position of the feature, with sequence numbering starting at 1.\n",
    "6. score - A numeric score.  Not important for this workshop\n",
    "7. strand - defined as + (forward) or - (reverse).\n",
    "8. frame - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on..\n",
    "9. **attribute** - A semicolon-separated list of tag-value pairs, providing additional information about each feature.\n",
    "\n",
    "Pay attention to the columns in bold.  We will work with these in this tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 2\n",
    "\n",
    "The `wc` (word count) command is very simple. It counts lines, words and characters in text. \n",
    "\n",
    "For example try it on the `h_pylori.gff` file\n",
    "\n",
    "```\n",
    "wc h_pylori.gff\n",
    "```\n",
    "\n",
    "The output includes (from left to right) the number of lines, words and characters.  \n",
    "\n",
    "Look at the help for `wc` and find the option for printing just the number of lines.\n",
    "```\n",
    "wc --help\n",
    "```\n",
    "\n",
    "**Your Task:** Write code to print the number of lines in the `h_pylori.gff` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e2_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e2_answer(){\n",
    "### BEGIN SOLUTION\n",
    "wc -l h_pylori.gff\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e2_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3\n",
    "\n",
    "Up to now we have always used just a single command but the real power of unix comes from the ability to combine commands.  One way to combine commands is by using the pipe operator, which is just a vertical bar `|`.  \n",
    "\n",
    "For example we could pipe the output of the `head` command into `wc` like this\n",
    "\n",
    "```\n",
    "head h_pylori.gff | wc\n",
    "```\n",
    "\n",
    "Notice that the `wc` command on the right doesn't need to specify an input file.  This is because it's input is the output from the previous command.  Not all commands can be used in pipes like this but most of the basic unix ones can. \n",
    "\n",
    "**Your task:** Write a command which prints the number of words in the top 5 lines of `h_pylori.gff`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e3_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e3_answer(){\n",
    "### BEGIN SOLUTION\n",
    "head -n 5 h_pylori.gff | wc -w\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e3_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 4\n",
    "\n",
    "The `grep` command is used to search for patterns in text.  For example we could search for all entries containing the word `pathogenicity` to find features related to pathogenicity.\n",
    "\n",
    "```bash\n",
    "grep 'pathogenicity' h_pylori.gff\n",
    "```\n",
    "\n",
    "The output is a bit messy because it includes a lot of information but if you read closely you should see that all these entries are related to the [cag pathogenicity island](https://www.ncbi.nlm.nih.gov/books/NBK2434/).  This is a group of genes that are thought to have been incorporated into the _H. pylori_ genome through lateral gene transfer (ie from another organism). Variations in this region are associated with _H. pylori_ infections that lead to gastric cancer.\n",
    "\n",
    "\n",
    "**Your task:** Write a command that outputs a count of the number of pathogenicity related features in the _H. pylori_ genome\n",
    "\n",
    "**Hint:** Each feature is on a separate line so counting lines is the same as counting features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e4_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e4_answer(){\n",
    "### BEGIN SOLUTION\n",
    "grep 'pathogenicity' h_pylori.gff | wc -l\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e4_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 5\n",
    "\n",
    "In exercise 3 we saw that a pipe can be used to send the output of one command to another.  Another thing we can do with output is send it to a file.  This is done using the output redirection operator `>`.\n",
    "\n",
    "For example we could create a subset of `h_pylori.gff` containing only pathogenicity related features like this\n",
    "\n",
    "```bash\n",
    "grep 'pathogenicity' h_pylori.gff > h_pylori_path.gff\n",
    "```\n",
    "\n",
    "Try running this command in a test cell.  Then try the `ls` command and you should see the new file that was created. \n",
    "\n",
    "A single operator like this, `>` will always overwrite the contents of the file where the output is sent.  \n",
    "\n",
    "A double operator like this `>>` will append to the end of the file.\n",
    "\n",
    "**Your task:** Use output redirection to create a file with the name `h_pylori_path_cds.gff` containing only entries that include the words 'pathogenicity' and 'CDS'\n",
    "\n",
    "**Note:** An easy way to grep for A **AND** B is to use two grep commands, piping output of the first to the second."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e5_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e5_answer(){\n",
    "### BEGIN SOLUTION\n",
    "grep 'pathogenicity' h_pylori.gff | grep 'CDS' > h_pylori_path_cds.gff\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e5_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 6\n",
    "\n",
    "Unix includes a powerful suite of tools for working with tabular data.  This includes [sort](https://en.wikipedia.org/wiki/Sort_&#40;Unix&#41;), [cut](https://en.wikipedia.org/wiki/Cut_&#40;Unix&#41;), [join](https://en.wikipedia.org/wiki/Join_&#40;Unix&#41;) as well as the much more general purpose tool [awk](https://en.wikipedia.org/wiki/Awk).\n",
    "\n",
    "The `cut` command can be used to extract columns from a file. \n",
    "\n",
    "For example we could extract just the first three columns from `h_pylori.bed` like this\n",
    "\n",
    "```bash\n",
    "cut -f 1,2,3 h_pylori.gff\n",
    "```\n",
    "\n",
    "Here we used `-f` to specify that we wanted to cut in `field` mode (ie column delimited as opposed to byte or character).  The most common way to use `cut` is with the `-f` option as this is the most natural.  We also used `1,2,3` to specify that we wanted columns 1, 2 and 3.  \n",
    "\n",
    "**Your task:** Write a command to extract only the start and end coordinates (columns 4 and 5) from pathogenicity related entries of `h_pylori.gff`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e6_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e6_answer(){\n",
    "### BEGIN SOLUTION\n",
    "grep 'pathogenicity' h_pylori.gff | cut -f 4,5\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e6_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sorting coordinates\n",
    "\n",
    "Another very useful command for use with tables is `sort`.  To get an idea of how the pathogenicity genes are spaced it might be useful to sort by their start coordinate.  Since the start coordinate is in column 4 we can sort as follows;\n",
    "\n",
    "```bash\n",
    "grep 'pathogenicity' h_pylori.gff | sort -n -k 4\n",
    "```\n",
    "\n",
    "To make this easier to read we `cut` out only columns 1-5\n",
    "\n",
    "```bash\n",
    "grep 'pathogenicity' h_pylori.gff | sort -n -k 4 | cut -f 1-5\n",
    "```\n",
    "\n",
    "Look closely at the start and end coordinates in the output from the above command (run it in a test cell).  You should be able to see that all of the genes occur together with relatively little space between them. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 7\n",
    "\n",
    "Another use of the `sort` command is to find all the unique entries in a column.  We can do this using the `-u` (unique) option as follows.  Remember that column 3 in `gff` is the feature type so this should show a list of all the possible feature types in the file.\n",
    "\n",
    "```bash\n",
    "cut -f 3 h_pylori.gff | sort -u\n",
    "```\n",
    "\n",
    "Notice that this results in a few items that begin with a `#` character.  These aren't useful because they aren't feature types at all.  They are comments which means that they should be ignored. \n",
    "\n",
    "One way to remove these would be to use the `-v` option of the grep command.  This allows us to use `grep` to search for text and **exclude** all lines containing that text.  It is essentially a way of inverting a search. \n",
    "\n",
    "```bash\n",
    "cut -f 3 h_pylori.gff | sort -u | grep '#' -v\n",
    "```\n",
    "\n",
    "**Your task:** Write a command to show all possible unique values for the first column of `h_pylori.gff`, excluding comments\n",
    "\n",
    "**Interpretation:** Given what you know about the `gff` format and the fact that _H. pylori_ is a Bacterium why is there only a single unique value in column 1 of the file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e7_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e7_answer(){\n",
    "### BEGIN SOLUTION\n",
    "cut -f 1 h_pylori.gff | sort -u | grep '#' -v\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e7_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e7"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Create Assignment",
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
