{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workshop 3 - Finding bacterial genes\n",
    "\n",
    "In this workshop you will use your `Bash` skills to translate bacterial DNA into amino acid sequences and use this to discover the location of protein coding genes.\n",
    "\n",
    "Before you begin this workshop you should have familiarised yourself with chapters 1-25 of the [interactive guide](http://bc2023.bioinformatics.guide/lessons/).\n",
    "\n",
    "At the end of this workshop you should be able to;\n",
    "\n",
    "1. Recognise and understand FASTA formatted data\n",
    "2. Use command line software to translate DNA in all 6 frames\n",
    "3. Identify candidate genes on the basis of long open reading frames\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IMPORTANT\n",
    "\n",
    "> ## Run the Setup Code \n",
    "\n",
    "In order for this notebook to work properly you need to **run the cell below before doing anything else**. This will load custom functions and settings required to make the self assessment exercises work. \n",
    "\n",
    "If you restart your kernel you will also need to rerun the setup code \n",
    "\n",
    "> ## Don't use the `cd` command \n",
    "\n",
    "The answers to all self assessment exercises assume that you don't change your directory from the default.  You shouldn't ever need to use the `cd` command to answer an exercise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "nbgrader": {
     "grade": false,
     "grade_id": "setup",
     "locked": true,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Setup Done\n"
     ]
    }
   ],
   "source": [
    "wget -q https://www.dropbox.com/s/es7lr8sy4ubg3ug/translate.py?dl=0 -O translate.py\n",
    "chmod u+x translate.py\n",
    "\n",
    "wget -q https://www.dropbox.com/s/4irxxomz3aajtou/setup.sh?dl=0 -O setup.sh\n",
    "source setup.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background\n",
    "\n",
    "<a title=\"Times 5 magnification of colonies of Burkholderia pseudomallei on Ashdown's agar after 4 days' incubation.. By Gavin Koh (Own work) [GFDL (http://www.gnu.org/copyleft/fdl.html) or CC BY-SA 4.0-3.0-2.5-2.0-1.0 (http://creativecommons.org/licenses/by-sa/4.0-3.0-2.5-2.0-1.0)], via Wikimedia Commons\" href=\"https://commons.wikimedia.org/wiki/File%3ABps_close.JPG\"><img width=\"256\" alt=\"Bps close\" src=\"https://upload.wikimedia.org/wikipedia/commons/e/ee/Bps_close.JPG\" style=\"float: right;margin-right: 7px;margin-top: 7px;\"/></a>\n",
    "\n",
    "Imagine that you are part of a team of researchers working on the pathogenic bacteria, [_Burkholderia pseudomallei_](http://en.wikipedia.org/wiki/Burkholderia_pseudomallei). _B. pseudomallei_ is a soil dwelling bacterium and is the causal agent of [melioidosis](http://en.wikipedia.org/wiki/Melioidosis)  \n",
    "\n",
    "You have sequenced a 20kb long fragment of the _B. pseudomallei_ genome.  Through a mutagenesis screen you know that this fragment contains a region that is implicated in eukaryotic cell invasion.  \n",
    "\n",
    "Your task is to map the locations of protein coding genes within this region and determine whether any of them overlap with the known mutant site.  We will complete this task over the course of tutorials 3 and 4."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTA format\n",
    "\n",
    "Use the `ls` command to list the files in your current directory.  One of the files should be called `genomic_sequence.fasta`. This file contains the 20kb long genome fragment in [FASTA](http://en.wikipedia.org/wiki/FASTA_format) format.\n",
    "\n",
    "FASTA is used for both nucleotide and amino acid sequences and is very popular due to its simplicity. A FASTA file consists of one or more sequence entries.  Each sequence entry begins with a description line which begins with a `>`.  Following the description line comes the actual sequence.  When another description line is encountered it denotes a separate entry.\n",
    "\n",
    "Inspect the top few lines of the file with the `head` command and verify that it is is in FASTA format.\n",
    "\n",
    "```bash\n",
    "head genomic_sequence.fasta\n",
    "```\n",
    "\n",
    "Since only the header lines contain the `>` symbol we can confirm that there is just one entry in the file by searching for these symbols with the `grep` command as follows;\n",
    "\n",
    "```bash\n",
    "grep '>' genomic_sequence.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 1\n",
    "\n",
    "Except for the first line, all of the text in the `genomic_sequence.fasta` file consists of DNA sequence.  Your goal in this exercise is to determine the length of the sequence (ie how many nucleotide bases are represented). \n",
    "\n",
    "The `wc` command can be use to count lines of text.  For example;\n",
    "\n",
    "```bash\n",
    "wc -l genomic_sequence.fasta\n",
    "```\n",
    "\n",
    "There are 80 bases per line and the last line contains 55 bases.\n",
    "\n",
    "**Your task:** Determine the length of the genomic sequence in `genomic_sequence.fasta`\n",
    "\n",
    "Use the `echo` command to make your answer work with the autograder. For example let's say your answer is `45` you would enter it as;\n",
    "\n",
    "```\n",
    "echo \"45\"\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e1_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e1_answer(){\n",
    "### BEGIN SOLUTION\n",
    "echo \"20455\"\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e1_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optional\n",
    "\n",
    "A better tool for obtaining FASTA sequence lengths (and much more) is [bioawk](https://github.com/lh3/bioawk) . This tool is beyond the scope of this tutorial but if you are curious you may wish to try it.\n",
    "\n",
    "The general form of the bioawk command for determining sequence lengths is shown below.  See if you can reformulate this command to give the answer to exercise 1.\n",
    "\n",
    "```bash\n",
    "bioawk -c fastx '{print length($seq)}' <input>.fasta\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In-silico Translation\n",
    "\n",
    "In principle the amino acid sequences of proteins encoded by bacterial genes can be determined simply by translating directly from the genomic sequence. Identifying the starting sites of translation in these genes is a little more challenging and will be explored in more detail later. \n",
    "\n",
    "Given the bacterial DNA sequence\n",
    "```\n",
    "GCTGCGTGACGTACG\n",
    "```\n",
    "We would translate in the first forward frame by first breaking the sequence into its codons.\n",
    "```\n",
    "GCT GCG TGA CGT ACG\n",
    "```\n",
    "And then using [the standard genetic code](http://en.wikipedia.org/wiki/DNA_codon_table) to translate these to amino acids. \n",
    "```\n",
    "GCT GCG TGA CGT ACG\n",
    " A   A   *   R   T\n",
    "```\n",
    "**Note:** A stop codon `*` indicates termination of translation.\n",
    "\n",
    "### Reading frames\n",
    "\n",
    "A DNA sequence can be translated in six possible ways depending on the direction and starting base of translation.\n",
    "\n",
    "**Forward**\n",
    "\n",
    "The example above shows translation based on the first forward reading frame. In this frame we read from left to right and start at the first position. If we were translating in the second forward reading frame we would obtain codons and translate as follows;\n",
    "```\n",
    "G CTG CGT GAC GTA CG\n",
    "   L   R   D   V\n",
    "```\n",
    "Note that we discarded one base from the left.  This also resulted in two discarded bases at the right but overall it resulted in a longer contiuous amino acid sequence because there was no stop codon.\n",
    "\n",
    "The third forward reading frame is obtained by discarding one more base from the left.\n",
    "```\n",
    "GC TGC GTG ACG TAC G\n",
    "    C   V   T   Y\n",
    "```\n",
    "\n",
    "What if we kept going? Since there are three bases in a codon, discarding three from the left is equivalent to the first forward frame.\n",
    "```\n",
    "GCT GCG TGA CGT ACG\n",
    "     A   *  R    T\n",
    "```\n",
    "\n",
    "**Reverse**\n",
    "\n",
    "Reverse frames correspond to translation on the opposite strand of DNA. The bases on the opposite strand will be complementary to the forward strand and will be read in reverse. We therefore translate based on the reverse complement. For our example sequence this is;\n",
    "```\n",
    "CGTACGTCACGCAGC\n",
    "```\n",
    "Translation in the first reverse frame is therefore\n",
    "```\n",
    "CGT ACG TCA CGC AGC\n",
    " R   T   S   R   S\n",
    "```\n",
    "The second and third reverse frames can be derived on the complementary sequence in exactly the same way as the forward frames on the forward sequence.\n",
    "\n",
    "### Software\n",
    "\n",
    "The tutorial directory provides a program `translate.py` that translates DNA in all 6 frames. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 2\n",
    "\n",
    "When working with large datasets it is often useful to run some test code on a small fraction of the data. \n",
    "\n",
    "The `head` command is useful here since it will extract the top `n` lines of a dataset where `n` can be any number we specify. \n",
    "\n",
    "For example to extract the top `10` lines of `genomic_sequence.fasta` we would do\n",
    "\n",
    "```bash\n",
    "head -n 10 genomic_sequence.fasta\n",
    "```\n",
    "\n",
    "Rather than creating a separate file with our test data we simply [pipe](https://en.wikipedia.org/wiki/Pipeline_&#40;Unix&#41;) it to other commands we want to run.  For example to verify that head did indeed extract only `10` lines we could pipe to `wc`.\n",
    "\n",
    "```bash\n",
    "head -n 10 genomic_sequence.fasta | wc -l\n",
    "```\n",
    "\n",
    "**Your task:** Translate the top 5 lines of `genomic_sequence.fasta` using translate.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e2_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e2_answer(){\n",
    "### BEGIN SOLUTION\n",
    "head -n 5 genomic_sequence.fasta | translate.py\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e2_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Familiarise yourself with `translate.py` and its options by doing the following;\n",
    "\n",
    "1. View the help.\n",
    "```bash\n",
    "translate.py -h\n",
    "```\n",
    "2. Translate the some sequence in the default frame (frame 1)\n",
    "```bash\n",
    "head genomic_sequence.fasta | translate.py\n",
    "```\n",
    "3. Translate in a different frame\n",
    "```bash\n",
    "head genomic_sequence.fasta | translate.py -f 2\n",
    "```\n",
    "4. Change output from fasta to tabular format\n",
    "```bash\n",
    "head genomic_sequence.fasta | translate.py -F 'tabular'\n",
    "```\n",
    "5. Split the translated sequence by stop codons\n",
    "```bash\n",
    "head genomic_sequence.fasta | translate.py -F 'tabular' -s\n",
    "```\n",
    "\n",
    "Don't worry if you don't understand all of the outputs at this stage. We'll cover that later.  For now just run the commands above and verify that the output roughly corresponds to the option you specified."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Distinguishing coding from non-coding sequences\n",
    "\n",
    "One of the ways we can distinguish real coding sequences from translated non-coding sequences is their length. \n",
    "\n",
    "Since `3/64` codons in the standard code are stops we would expect a stop codon to occur around 1 in every 21 amino acids if the underlying DNA sequence were random.  Based on this logic, the probability that a continuous sequence of 100 amino acids with no stop codons could be obtained from a random sequence of nucleotides is;\n",
    "\n",
    "$$p = (1-(3/64))^{100}$$\n",
    "$$p \\approx 0.01$$\n",
    "\n",
    "In other words, any sequences we obtain that are $100$ amino acids or longer are unlikely to be non-coding sequences.\n",
    "\n",
    "**Note:** Although it is a useful first approximation to assume that non-coding sequence is random this is not strictly true. Non-coding sequences are unlikely to be truly random because they are often functional (eg regulatory) or may be derived from sequences that were once coding (eg pseudogenes).  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 3\n",
    "\n",
    "Use `translate.py` to translate `genomic_sequence.fasta` in the first frame\n",
    "\n",
    "```bash\n",
    "translate.py -f 1 genomic_sequence.fasta\n",
    "```\n",
    "\n",
    "The `tr` command can be used to delete all characters other than a specified character. For example to output only `A`'s in the sequence we would do;\n",
    "\n",
    "```bash\n",
    "translate.py -f 1 genomic_sequence.fasta | tr -cd 'A' \n",
    "```\n",
    "\n",
    "The number of `A`'s can then be counted with\n",
    "\n",
    "```bash\n",
    "translate.py -f 1 genomic_sequence.fasta | tr -cd 'A' | wc -m\n",
    "```\n",
    "\n",
    "**Your task:** Write a command to output the number of stop codons in the frame 1 translation of `genomic_sequence.fasta`\n",
    "\n",
    "Compare your answer with the expected number of stop codons if the sequence was random.  What does this tell you about the gene content of our sequence in frame 1?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e3_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e3_answer(){\n",
    "### BEGIN SOLUTION\n",
    "translate.py -f 1 genomic_sequence.fasta | tr -cd '*' | wc -m\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e3_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 4\n",
    "\n",
    "Modify the command pipeline you created for exercise 3 to count amino acids.\n",
    "\n",
    "Remember that the header line `>NC_006351_f_1` contains some letters that could be confused with amino acids (ie `N` and `C`). \n",
    "\n",
    "You can deal with this by using the `tail` command to remove the first line of output like this\n",
    "\n",
    "```bash\n",
    "translate.py -f 1 genomic_sequence.fasta | tail -n +2 \n",
    "```\n",
    "\n",
    "**Your task:** Write code to output the number of cysteines in the frame 1 translation of `genomic_sequence.fasta`\n",
    "\n",
    "Try your code on various other amino acids.  Do they vary much in abundance?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e4_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e4_answer(){\n",
    "### BEGIN SOLUTION\n",
    "translate.py -f 1 genomic_sequence.fasta | tail -n +2 | tr -cd 'C' | wc -m\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e4_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Open Reading Frames\n",
    "\n",
    "For the purposes of this tutorial we will define an open reading frame as the part of a reading frame that [does not contain stop codons](https://www.nature.com/subjects/open-reading-frames).  Some definitions also require that an ORF begin with a start codon.\n",
    "\n",
    "Our strategy for finding genes in the _B. mallei_ genome will be to first look for long ORFs as these are unlikely to occur in non-coding regions.\n",
    "\n",
    "So far we have used `translate.py` in its default output mode which simply translates the genome into one long amino acid sequence, including stop codons.  To make finding long orfs easier we will use the `--split` option to divide up the translation into its constituent ORFs. Eg;\n",
    "\n",
    "```bash\n",
    "translate.py -f 1 --split genomic_sequence.fasta | head\n",
    "```\n",
    "\n",
    "To make things even easier we change the output format to `tabular`.  This is very useful because unix provides many ways to manipulate tabular files\n",
    "\n",
    "```bash\n",
    "translate.py -f 1 --split -F 'tabular' genomic_sequence.fasta | head\n",
    "```\n",
    "\n",
    "The tabular format consists of the following fields (columns)\n",
    "\n",
    "```bash\n",
    "parent_id    start   end   strand   orf_number   orf_length   orf_sequence\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 5\n",
    "\n",
    "The `translate.py` script usually works with one frame at a time but you can ask it to translate in all 6 frames by providing `0` for the `frame` argument. ie\n",
    "\n",
    "```bash\n",
    "translate.py -f 0 --split -F 'tabular' genomic_sequence.fasta\n",
    "```\n",
    "\n",
    "**Your Task:** Write code to count the number of ORFs generated by translating `genomic_sequence.fasta` in all 6 frames.  Your answer should produce output in the form of a single number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e5_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e5_answer(){\n",
    "### BEGIN SOLUTION\n",
    "translate.py -f 0 --split -F 'tabular' genomic_sequence.fasta | wc -l\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e5_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sorting tabular output\n",
    "\n",
    "The `sort` command is very useful when working with tabular files.\n",
    "\n",
    "Have a look at the help for sort\n",
    "```bash\n",
    "sort --help\n",
    "```\n",
    "As you can see it has many options.  For our purposes the most useful options are\n",
    "```\n",
    "-k, --key\n",
    "-n, --numeric-sort\n",
    "```\n",
    "\n",
    "The `k` option allows us to specify a column to sort on. For example we could sort according to the 3rd column with\n",
    "\n",
    "```bash\n",
    "sort -k 3\n",
    "```\n",
    "\n",
    "By default, `sort` sorts alphabetically. Providing the `n` option changes this behaviour to a numeric sort (which we often want). By combining `k` and `n` options we could do a numeric sort on the 3rd field\n",
    "\n",
    "```bash\n",
    "translate.py -f 1 --split -F 'tabular' genomic_sequence.fasta | sort -n -k 3 | head\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 6\n",
    "\n",
    "In this exercise we will finally obtain a list of candidate ORFs that may represent bacterial gene products.\n",
    "\n",
    "**Preparation:** \n",
    "\n",
    "1. Use a test cell to formulate a command that prints all ORFs (in all frames) (in tabular format) sorted by length.\n",
    "2. Use the `tail` command to display only the longest of these ORFs\n",
    "3. Use the `-n` option to `tail` to adjust the number of long ORFs displayed.  \n",
    "4. Adjust the displayed number of lines until you only display ORFs equal to or longer than `100`.\n",
    "5. Use `grep` to search for methionines `M`.\n",
    "6. Use `wc -l` to produce a count.\n",
    "\n",
    "**Your task:** Write code to print the number of ORFs from a 6 frame translation of `genomic_sequence.fasta` that have length greater than or equal to 100 **AND** contain at least one start codon `M`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true,
    "nbgrader": {
     "grade": false,
     "grade_id": "e6_answer",
     "locked": false,
     "schema_version": 1,
     "solution": true
    }
   },
   "outputs": [],
   "source": [
    "e6_answer(){\n",
    "### BEGIN SOLUTION\n",
    "translate.py -f 0 --split -F 'tabular' genomic_sequence.fasta | sort -n -k 6 | tail -n 100 | grep 'M' | wc -l\n",
    "### END SOLUTION\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "nbgrader": {
     "grade": true,
     "grade_id": "e6_test",
     "locked": true,
     "points": 1,
     "schema_version": 1,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your answer is correct\n"
     ]
    }
   ],
   "source": [
    "test_e6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "\n",
    "In this workshop we formulated a command to translate a genomic sequence from _Burkholderia pseudomallei_ in all $6$ reading frames, and extract a list of genomic regions corresponding to candidate genes (exercise 6).\n",
    "\n",
    "This process forms a good starting point for a gene prediction algorithm.  Nevertheless, it is likely that many of the sequences in our list of long open reading frames are not actually transcribed by the organism.  To refine the list further we could;\n",
    "\n",
    "1. Identify other features of the genomic sequence that characterise true genes. For example translation start sites are generally accompanied by a ribosome binding sequence $~17bp$ upstream.\n",
    "2. Use a program such as [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) to look for homologous sequences that have been identified as transcribed genes in other organisms\n",
    "3. Use experimental evidence (RNA sequencing or Mass Spectrometry) to verify expression.\n",
    "\n",
    "These steps (both computational and experimental) form the basic workflow of most modern genome annotation projects. Finding better ways to describe the functional elements of genomes (eg genes) is a significant challenge and an active area of research."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A note about `awk`\n",
    "\n",
    "A better way to complete exercise 6 would be to use a tool called [awk](https://en.wikipedia.org/wiki/AWK). `AWK` is an old but very powerful and versatile tool but can be challenging for beginners.  We will cover a little of its functionality here but [this tutorial](http://www.grymoire.com/Unix/Awk.html) is a good introduction if you would like to learn more.\n",
    "\n",
    "In workshop 4 you will need to use some very basic features of `awk`.  As an introduction, let's consider how exercise 6 could have been accomplished with `awk`\n",
    "\n",
    "The general syntax of `awk` is\n",
    "```\n",
    "awk '<program>' <file>\n",
    "```\n",
    "where `<file>` is the name of a text file we want to process, and `<program>` is an `awk` program which is run on each line of the text file. `<file>` can also be data piped from a previous command, in which case the syntax would be;\n",
    "```\n",
    "other_command | awk '<program>'\n",
    "```\n",
    "\n",
    "The general structure of an `awk` program is;\n",
    "```\n",
    "condition { action }\n",
    "```\n",
    "\n",
    "In exercise 6 we had to play around with `sort` and `tail` to figure out how many ORFs were `>=` 100 amino acids long.  With `awk` we can simply solve this by using a condition, that the number in column 6 (the length) is greater than or equal to `100`. So our command would be\n",
    "\n",
    "```\n",
    "translate.py -f 0 --split -F 'tabular' genomic_sequence.fasta | awk '$6>=100{print $0}'\n",
    "```\n",
    "\n",
    "In an `awk` program each line is separated into fields (ie columns) which can be referenced by a `$` followed by the field number.  So column `6` would be `$6`.\n",
    "\n",
    "The `condition` here is;\n",
    "```\n",
    "$6>=100\n",
    "```\n",
    "which means that the `action` will only be run if the value in column `6` is greater than or equal to `100`.\n",
    "\n",
    "The `action` is;\n",
    "```\n",
    "print $0\n",
    "```\n",
    "which means print the entire line because `$0` is shorthand for the entire line.\n",
    "\n",
    "We can insert another condition to check for `M`'s in the sequence field.\n",
    "\n",
    "```\n",
    "translate.py -f 0 --split -F 'tabular' genomic_sequence.fasta | awk '$6>=100 && $7~/M/{print $0}'\n",
    "```\n",
    "\n",
    "The answer to exercise 6 can then be obtained by piping to `wc`\n"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Create Assignment",
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
